src/core/Observable_stat.hpp

```diff
@@ -19,59 +19,220 @@
 #ifndef ESPRESSO_OBSERVABLE_STAT_HPP
 #define ESPRESSO_OBSERVABLE_STAT_HPP
 
-#include <vector>
+#include <boost/range/numeric.hpp>
 
-struct Observable_stat {
-  /** Status flag for observable calculation. For 'analyze energy': 0
-   *  re-initialize observable struct, else everything is fine,
-   *  calculation can start. For 'analyze pressure' and 'analyze p_inst':
-   *  0 or !(1+v_comp) re-initialize, else all OK.
-   */
-  int init_status;
+#include <utils/Span.hpp>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
 
+/** Cache for system statistics.
+ *  Store unidimensional (energy, scalar pressure) and multi-dimensional
+ *  (stress tensors) properties of the system and provide accumulation and
+ *  reduction functionality.
+ */
+class Observable_stat_base {
+protected:
   /** Array for observables on each node. */
   std::vector<double> data;
+  /** Number of doubles per data item */
+  size_t m_chunk_size;
+
+public:
+  /** @param chunk_size Dimensionality of the data
+   */
+  explicit Observable_stat_base(size_t chunk_size) : m_chunk_size(chunk_size) {}
+
+  /** Resize the observable */
+  virtual void resize() = 0;
+
+  /** Resize and zero out the observable */
+  void resize_and_clear() {
+    resize();
+    std::fill(data.begin(), data.end(), 0);
+  }
+
+  /** Gather the contributions from the current MPI rank.
+   *  @param[out] out Destination of the reduction.
+   */
+  void reduce(double *out) const;
+
+  /** Accumulate values.
+   *  @param acc    Initial value for the accumulator.
+   *  @param column Which column to sum up (only relevant for multi-dimensional
+   *                observables).
+   */
+  double accumulate(double acc = 0.0, size_t column = 0) {
+    assert(column < m_chunk_size);
+    if (m_chunk_size == 1)
+      return boost::accumulate(data, acc);
+
+    for (auto it = data.begin() + column; it < data.end(); it += m_chunk_size)
+      acc += *it;
+    return acc;
+  }
+
+protected:
+  /** Get contribution from a non-bonded interaction */
+  Utils::Span<double> non_bonded_contribution(double *base_pointer, int type1,
+                                              int type2) const {
+    extern int max_seen_particle_type;
+    if (type1 > type2) {
+      using std::swap;
+      swap(type1, type2);
+    }
+    int offset = ((2 * max_seen_particle_type - 1 - type1) * type1) / 2 + type2;
```

Comment:
This is basically what `Utils::upper_triangular` does I think?

Reply:
indeed, thanks!